### DISCUSSION TITLE
[QUESTION] Any instructions about working with geometric data needed

### DISCUSSION THREAD

**Discussion by pndvg** [idea-statica/ideastatica-public#145] [General Discussion]
**This is forwarded request (00023342) to technical support.**
Hello IS Team!
Is there any available methods for setting the parameters of connection, available in Idea Statica Connection user interface? We want to vary with C# and Excel some parameters (especially geometric & cross sections), visible in user interface (like on the picture below). I have already looked through the examples, and haven't found any related to my case. Maybe these geometric parameters of the connections shall be somehow prepared in IS app before using the template with API? Actually this is only one obstackle stopping the work on app. Hope to receive your comments soon.
![Idea](https://user-images.githubusercontent.com/67465772/154024554-97186f0b-a61f-48a3-832a-b05c0fe5ac6c.png)

**About the problem in common:**
I'm working on app, which should work on the top of IdeaStatica. It should calculate series of connection types using predefined loads and geometric parameters. The main idea(not Statica)) : app opens connection template (ok) -> using excel table app applies loads, codes, specifies the geometries and other initial data (**not ok**) -> creates calculation files in IS (ok) -> calculates connection(ok) -> gather raw results (ok) -> pack them into excel tables (ok).

**Comment by nathanlukeIDEA** [comment]
Hi Pavel, 

Thanks for your question. 

We do have some availability to set parameters in a defined IDEA Connection file. These are parameters available in operation dialogues. You will need to find the appropriate parameters which relate to the dimensions you are highlighting in your connection.

You can use the instructions at this link to enable User Mode 16 which will provide access to a Parameters Tab: https://github.com/idea-statica/ideastatica-public/blob/main/docs/params-in-ideacon-project.md

Below is shown how you can create a string parameter that is set to the BP1 Operation offset. 

![image](https://user-images.githubusercontent.com/93578548/154593772-1fe72c16-e3d9-4dc3-ba77-1f7e671b6737.png)

Steps are:
1.	Add parameter with '+' button next to parameters (on left).
2.	Name the parameter in Parameter Id. Then keep selected.
3.	Click '+' button next to 'Parameters in model' and scroll to the desired model parameter you would like to relate to the created parameter. In this instance the BP1 operation offset parameter.
 
Once you have defined the parameters, you can use c# to manipulate these using the `ConnectionHiddenCheckClient`. 

Steps below:
1.	Open Connection:` client.OpenProject(IdeaConFileName)`;
2.	Call the `client.GetParametersJSON(conInfo.Identifier) `to retrieve the JSON string of the parameters you have defined. 
3.	You should deserialize this string into an Array of Parameters (to enable updating in code)
4.	Set your updated values (from your excel file) through code by name/value reference
5.	Serialize the Parameter Array back to a JSON string.
6.	Call `client.ApplyParameters(conIdentifier, updatedParameters)`; 
7.	Save the updated file using `client.SaveAsProject(filepath)` Note: this needs to be a different file name to the open project. We are improving this in one of the upcoming releases to enable a direct `Save()` method.

Loads can be updated in a similar manner as explained for parameters. Call the `client.GetConnectionLoadingJSON(conInfo.Identifier)` to retrieve the JSON string of the loading.  Use similar serialize/deserialize steps as above and then call `client.UpdateLoadingFromJson(conIdentifier, updateLoading)`

I have attached an example C# file that will help you with this. It is slightly more complex than what you are trying to achieve (it is a parameter optimization example), but it steps through updating parameters. You will need to add the IdeaStatiCa.PlugIn package to your Visual Studio Project.

Let me know if this does not help, or if there are further limitations to what you are trying to achieve. 

[IdeaParameterOptimization.zip](https://github.com/idea-statica/ideastatica-public/files/8093135/IdeaParameterOptimization.zip)

**Comment by pndvg** [comment]
Hi,
I have already checked the possibilities of parametrisation - thats what I need. Few simple questions in addition of the topic:
1. Where to find these GUID's for the profile tables? Default values of GUID's work only with HEA-catalog. Maybe you have public dictionary somewhere? 
![image](https://user-images.githubusercontent.com/67465772/154686976-476cc3e4-ec21-43fc-b9ec-d213310019b2.png)
2. Which way bolts shall be parametrised? Mean: diameter, type, rows count/steps.
3. Which format is preferrable for "template with parameters inside" storing: IOM(XML), standard *.ideaCon or *.contemp? Actually I have no idea, why Idea use these three, since Idea has "Template" and "Project" types of files for two scenarious. Other seems to be covered with API. Maybe you can clarify this questin in addition to your recommendation. 

Sincerely Yours,
PN

**Comment by nathanlukeIDEA** [comment]
Hi Pavel, 

Some responses to your numbered items. Please note that there are current limitations with parameters that we plan on improving. It is good to see the parameters interface being useful, and your feedback allows us to start improving it. 

1.	We are aware of this issue. The TableId and ItemId refer to values in the MPRL and are not easily referenced. These parameters are likely to be removed soon and replaced with references to the ‘project’ library.

2.	Generally, updating single parameter value items through the available ‘model parameters’ for each operation is the simplest approach (although it is currently possible to define an entire anchor parameter). The best way to view all available model parameters is to create a new ‘expression’ parameter. No filtering of model parameters will be applied, and you will see all available parameters relating to anchors in an operation. It would be good to get some further feedback on whether this explained approach is preferred or not. Again, this should be improved as currently, the Bolt Assemblies reference an MPRL item, not a project item.

3.	Currently parameters are only stored in the IDEA Connection Project (.ideaCon).
* The .contemp (.XML) file stores project operation data that can then be applied to another project with similar member geometry defined. As you point out there are API calls that can allow you to apply these template files to other projects. These calls are typically used where many similar connections are created through IOM (with only member geometry and loads). The template can be applied without the need to open the connection. 
* IOM is the file used when creating a connection file from a third-party application. It can transfer basically any geometry but does not transfer any ‘smart’ operations or the project parameters you are defining.  I.e. all welds are defined separately as single welds, not as a part of an operation.

It would be good to discuss some of these items further detail over a video call.

Cheers
Nathan

**Comment by nathanlukeIDEA** [comment]
@PavelNedviga In version 23 of IDEA StatiCa, the parameters interface has been improved significantly. You should now be able to define all anchor geometry relatively easily. Also changing anchor sizes is also possible.

https://www.ideastatica.com/support-center/setting-up-parametric-design-with-developer-tab

**Reply by pndvg** [reply]
Hi! Sounds great! Unfortunately, we can't extend our subscription since the company is located in Russia(. Hope the situation will change. Anyway, I'm looking forward to checking the new feature!
